### PR TITLE
Gui/packet parser tool

### PR DESCRIPTION
**Pull request by chayaRoten**
**Purpose of the Code**

The purpose of this code is to provide a versatile and robust tool for parsing and interpreting packet data based on a specified JSON format. This tool is designed to handle various types of data fields and bit fields, allowing for flexible and efficient extraction of information from raw data buffers. The primary goal is to enable accurate and customizable data parsing, suitable for a range of applications that involve complex data structures.

**Key Features:**

**PacketParser Class:**

Overview: The PacketParser class is central to this implementation. It is designed to read and interpret data packets based on a JSON configuration file that defines the structure and types of fields.
Constructor: The constructor of the PacketParser class takes a file path to a JSON configuration and a pointer to a raw data buffer. It initializes the parser with the provided configuration and data.
Field Access: The class provides methods to retrieve values for specific fields and to get all field values from the data buffer. This allows users to extract and inspect data from packets in a structured manner.
Field Types: It supports a variety of field types, including unsigned integers, signed integers, character arrays, floating-point numbers (fixed and mantissa formats), doubles, and booleans.
Bit Field Handling: The class can handle bit fields, which are complex fields that consist of multiple sub-fields. It supports the extraction and decoding of these sub-fields according to their definitions in the JSON configuration.
Endianness: The class takes into account the byte order (endianness) specified in the JSON file, which can be either "little" or "big". This ensures that data is interpreted correctly based on the specified endianness.
Error Handling and Validation: The PacketParser includes methods for validating field sizes and ensuring that data conforms to the expected formats. This helps prevent errors and ensures that parsed data is accurate.

**JSON Configuration File:**

Purpose: The JSON configuration file defines the structure and types of fields in the data packets. It specifies details such as field names, types, sizes, and nested bit field structures.
Structure: The JSON file includes descriptions and constraints for each field type and provides examples of how fields and bit fields should be defined. This configuration file serves as a blueprint for how the PacketParser interprets the raw data buffer.
Field Definitions: The JSON allows for the definition of various field types, including character arrays, floating-point numbers, and bit fields with nested sub-fields. It also specifies the size of each field and provides rules for valid field sizes.

**Testing:**

Objective: Tests have been created to ensure the functionality and accuracy of the PacketParser class. These tests verify that the class correctly parses and interprets different types of data fields and bit fields.
Scope: The tests include cases from the image processing team’s codebase, ensuring that the parser handles real-world data scenarios effectively.

**Summary**
In summary, this code introduces a generic and flexible packet parsing tool that leverages JSON configuration for defining data structures. The PacketParser class provides methods to extract and interpret various types of data from raw buffers, handling both simple and complex field types. The included JSON file format allows for detailed configuration of fields and their properties, while the comprehensive tests ensure the robustness and reliability of the parsing logic.

This implementation is valuable for applications that need to process and interpret structured data packets, offering a customizable and efficient solution for data parsing tasks.

### CODE REVIEW

File: parser_json/README.md
Comment (ChedvaErblich) [inline parser_json/README.md:1]: If you don't use it remove it.

File: parser_json/json_example.json [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      "description": "An array of field definitions. Each field has the following properties:",
+      "properties": {
+        "name": {
+          "description": "The name of the field. This is a string that will be used to identify the field.",
+          "example": "CharArrayField"
+        },
+        "type": {
+          "description": "The type of the field. Must be one of the following:",
+          "allowed_values": [
+            "char_array",
```
Comment (ChedvaErblich) [inline parser_json/json_example.json:19]: I don't see here the bool option

File: parser_json/src/packet_parser.cpp [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <nlohmann/json.hpp>
+#include <sstream>
+#include <stdexcept>
+#include "packet_parser.h"
+
+using json = nlohmann::json;
+
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.cpp]: Blank line

File: parser_json/src/packet_parser.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+   public:
+    PacketParser(const std::string &jsonFilePath);
+    void *getFieldValue(const std::string &fieldName) const;
+    std::map<std::string, FieldValue> getAllFieldValues() const;
+    const std::vector<Field> &getFields() const
+    {
+        return fields;
+    }
+    const void* getBuffer() const
+{
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.h]: Missing tab

File: parser_json/src/packet_parser.h [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    size_t offset;
+    std::vector<Field> fields;
+};
+
+class PacketParser {
+   public:
+    PacketParser(const std::string &jsonFilePath);
+    void *getFieldValue(const std::string &fieldName) const;
+    std::map<std::string, FieldValue> getAllFieldValues() const;
+    const std::vector<Field> &getFields() const
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.h]: It’s better to implement the functions in the .cpp file.

File: parser_json/src/packet_parser.h [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    void loadJson(const std::string &jsonFilePath);
+    uint32_t decodeUnsignedInt(const uint8_t *data, size_t bitLength) const;
+    int32_t decodeSignedInt(const uint8_t *data, size_t bitLength) const;
+    std::string decodeCharArray(const uint8_t *data, size_t byteLength) const;
+    float decodeFloat(const uint8_t *data, size_t bitLength) const;
+    double decodeDouble(const uint8_t *data, size_t bitLength) const;
+    bool decodeBoolean(const uint8_t *data, size_t bitLength) const;
+    uint8_t *extractBits(const uint8_t *bitFieldData, size_t offset,
+                         size_t size) const;
+    friend class PacketParserTest_ExtractBits_Test;
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.h]: It's better to avoid using friend.
For testing purpose it's o.k if you don't have other option,
So take a minute to think if you do it without friend.

File: parser_json/src/packet_parser.cpp [hunk 6]
```diff
@@ -0,0 +1,560 @@
...
+        }
+
+        currentOffset += byteLength;
+    }
+}
+
+uint32_t PacketParser::decodeUnsignedInt(const uint8_t *data,
+                                         size_t bitLength) const
+{
+    size_t byteLength = (bitLength + 7) / 8;  // Convert bits to bytes
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.cpp]: I see this conversion is used in multiple functions. 
Please extract it to a MACRO or a static function (try first with macro).

File: parser_json/src/packet_parser.cpp [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+#include <sstream>
+#include <stdexcept>
+#include "packet_parser.h"
+
+using json = nlohmann::json;
+
+
+FieldType PacketParser::getFieldType(const std::string &typeStr) const
+{
+    if (typeStr == "unsigned_int")
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.cpp]: Consider replacing it with a map that stores string-to-FieldType mappings. This will simplify the code, improve readability, and increase efficiency.

File: parser_json/src/packet_parser.cpp [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+        return FieldType::BOOLEAN;
+
+    return FieldType::UNKNOWN;
+}
+
+PacketParser::PacketParser(const std::string &jsonFilePath)
+{
+    loadJson(jsonFilePath);
+    setBuffer(nullptr);
+    std::cout << "Buffer after initialization in constructor: " << _buffer << std::endl;
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.cpp]: Consider using the logger.

File: parser_json/src/packet_parser.cpp [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+    if (type == "signed_int" && bitLength != 32) {
+        throw std::runtime_error("Signed integers must be 32 bits.");
+    }
+    if (type == "boolean" && bitLength != 8) {
+        throw std::runtime_error("Boolean types must be 8 bits.");
+    }
+    if (type == "double" && bitLength != 64) {
+        throw std::runtime_error("Double types must be 64 bits.");
+    }
+    // Add more validations if needed
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.cpp]: If you have more validations please add,
add remove the comment...

File: parser_json/test/test_packet_parser.cpp [hunk 10]
```diff
@@ -0,0 +1,787 @@
...
+        parser.extractBits(buffer.data(), 8, 8);  // 8 bits from position 8
+    EXPECT_EQ(*extractedBits, 0xCD);              // Should be 11001101
+    std::free(extractedBits);
+
+    extractedBits =
+        parser.extractBits(buffer.data(), 16, 16);  // 16 bits from position 16
+    EXPECT_EQ(*extractedBits, 0xEF);                // Should be 11101111
+    std::free(extractedBits);
+
+    // ניקוי JSON שנוצר
```
Comment (ChedvaErblich) [inline parser_json/test/test_packet_parser.cpp]: Write your comments in English

File: parser_json/src/packet_parser.cpp [hunk 11]
```diff
@@ -0,0 +1,560 @@
...
+}
+
+uint8_t *PacketParser::extractBits(const uint8_t *bitFieldData, size_t offset,
+                                   size_t size) const
+{
+    // Calculate the number of bytes required to store the result
+    size_t numBytes = (size + 7) / 8;
+
+    // Allocate memory for the result
+    uint8_t *result = (uint8_t *)std::malloc(numBytes);
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.cpp]: Who is freeing?

File: parser_json/src/packet_parser.cpp [hunk 12]
```diff
@@ -0,0 +1,560 @@
...
+    if (!result) {
+        return nullptr;  // Allocation failed
+    }
+    std::memset(result, 0, numBytes);  // Initialize with zeros
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.cpp]: You can use calloc instead of malloc+memset.

File: parser_json/src/packet_parser.h [hunk 13]
```diff
@@ -0,0 +1,75 @@
+#ifndef PACKET_PARSER_H
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.h]: The convention is __PACKET_PARSER_H__

File: parser_json/src/packet_parser.cpp [hunk 14]
```diff
@@ -0,0 +1,560 @@
...
+        throw std::runtime_error("Double types must be 64 bits.");
+    }
+    // Add more validations if needed
+}
+
+void PacketParser::loadJson(const std::string &jsonFilePath)
+{
+    std::ifstream jsonFile(jsonFilePath);
+    if (!jsonFile.is_open()) {
+        throw std::runtime_error("Unable to open JSON file");
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.cpp]: Consider add more information, such as the file name etc...

File: parser_json/test/test_packet_parser.cpp [hunk 15]
```diff
@@ -0,0 +1,787 @@
...
+        try {
+            parser.getFieldValue("some_field"); // Attempt to access a field
+        } catch (const std::runtime_error &e) {
+            EXPECT_STREQ(e.what(), "Buffer is null");
+            throw;  // Re-throw to let the test framework handle the failure
+        }
+    }, std::runtime_error);
+
+    // Clean up the generated JSON file
+    cleanupJsonFile({jsonFilePath});
```
Comment (ChedvaErblich) [inline parser_json/test/test_packet_parser.cpp:891]: Add a test that init the parser and use it on 2 different buffers, by using setBuffer twice.

File: parser_json/src/packet_parser.cpp [hunk 16]
```diff
@@ -0,0 +1,560 @@
...
+        [&fieldName](const Field &field) { return field.name == fieldName; });
+
+    if (it != fields.end() && it->type != "bit_field") {
+        const Field &field = *it;
+        const uint8_t *data =
+            static_cast<const uint8_t *>(_buffer) + field.offset;
+        switch (getFieldType(field.type)) {
+            case FieldType::UNSIGNED_INT: {
+                uint32_t *value =
+                    new uint32_t(decodeUnsignedInt(data, field.size));
```
Comment (ChedvaErblich) [inline parser_json/src/packet_parser.cpp]: Let's discuss how we can avoid dynamic allocations, and if it's not possible, what's the best way to handle them.

File: parser_json/README.md
Comment (chayaRoten) [reply parser_json/README.md:1]: Done

File: parser_json/json_example.json (hunk 1 above)
Comment (chayaRoten) [reply parser_json/json_example.json:19]: Done

File: parser_json/src/packet_parser.cpp (hunk 2 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.cpp]: Done

File: parser_json/test/test_packet_parser.cpp (hunk 15 above)
Comment (chayaRoten) [reply parser_json/test/test_packet_parser.cpp:891]: Done

File: parser_json/src/packet_parser.cpp (hunk 16 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.cpp]: Done

File: parser_json/src/packet_parser.cpp (hunk 14 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.cpp]: Done

File: parser_json/src/packet_parser.h (hunk 13 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.h]: Done

File: parser_json/src/packet_parser.cpp (hunk 12 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.cpp]: Done

File: parser_json/src/packet_parser.h (hunk 4 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.h]: Done

File: parser_json/src/packet_parser.h (hunk 3 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.h]: Done

File: parser_json/src/packet_parser.h (hunk 5 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.h]: Done

File: parser_json/src/packet_parser.cpp (hunk 6 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.cpp]: Done

File: parser_json/src/packet_parser.cpp (hunk 7 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.cpp]: Done

File: parser_json/src/packet_parser.cpp (hunk 9 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.cpp]: Done

File: parser_json/test/test_packet_parser.cpp (hunk 10 above)
Comment (chayaRoten) [reply parser_json/test/test_packet_parser.cpp]: Done

File: parser_json/src/packet_parser.cpp (hunk 8 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.cpp]: Done

File: parser_json/src/packet_parser.cpp (hunk 11 above)
Comment (chayaRoten) [reply parser_json/src/packet_parser.cpp]: Done
